### ISSUE TITLE
`upload-file --sha1` option is documented poorly, consistently ignore, basically unusable

### ISSUE THREAD

**Issue by molomby** [Backblaze/B2_Command_Line_Tool#621]
For me (and I assume many of your other users) the ability to verify the integrity of data stored on your platform is critical. Yet the checksum mechanism that should provide this assurance seems to be unreliable and unintuitive.

The only docs I can find for the `--sha1` option come from `b2 help upload-file`:

> By default, upload_file will compute the sha1 checksum of the file to be uploaded.  But, if you already have it, you can provide it on the command line to save a little time.

On the surface this implies a few things:

* A sha1 checksum is always stored
* Providing a sha will cause the platform to store the supplied value
* Providing a sha will prevent the client software from calculating it's own

As far as I can tell, none of these are true.

## Not Computed By Default

Say I have a "big" file...

```sh
$ du -h ./big.zip
271M	./big.zip
```

Uploading `big.zip` doesn't cause a checksum to be computed or stored:

```sh
$ b2 upload-file molomby big.zip big.zip
big.zip: 100%|████████████████| 284M/284M [02:30<00:00, 1.89MB/s]
URL by file name: https://f002.backblazeb2.com/file/molomby/big.zip
URL by fileId: https://f002.backblazeb2.com/b2api/v2/b2_download_file_by_id?fileId=4_z80c47a5d0d48e29c7f100612_f2073f7e9b565b63b_d20200518_m024622_c002_v0001135_t0053
{
  "action": "upload", 
  "fileId": "4_z80c47a5d0d48e29c7f100612_f2073f7e9b565b63b_d20200518_m024622_c002_v0001135_t0053", 
  "fileName": "big.zip", 
  "size": 284056801, 
  "uploadTimestamp": 1589769982000
}
$ b2 get-file-info 4_z80c47a5d0d48e29c7f100612_f2073f7e9b565b63b_d20200518_m024622_c002_v0001135_t0053 | jq '.contentSha1'
"none"
```

Apparently this is actually [by design](https://github.com/Backblaze/B2_Command_Line_Tool/issues/539#issuecomment-450705810)  and a result of the file being bigger than some arbitrary limit. The docs discuss this limit (`minPartSize`?) in terms of upload speed but don't mention its effect on functionality intended to protect data integrity. As a user this is confusing, unexpected and almost certainly not desirable.

But that's ok, I can generate my own hash and supply it, right?

## Not Stored When Supplied

```sh
$ shasum ./big.zip
c05a25f17b50a52944e44705ef8d7e40f47307ee  ./big.zip
$ b2 upload-file ---sha1 c05a25f17b50a52944e44705ef8d7e40f47307ee molomby big.zip big.zip
$ b2 get-file-info 4_z80c47a5d0d48e29c7f100612_f2138a7c2870a1c72_d20200518_m030637_c002_v0001138_t0021 | jq '.contentSha1'
"none"
```

Nope. Looks like the value I'm supplying is ignored on upload. What's more I can't see anyway the the `contentSha1` value can be set after the fact using this tool.

Well at least it works for small files, right?

## Supplied Value Ignored

```sh
$ du -h ./small.zip
 13M	./small.zip
$ shasum ./small.zip 
5f7270e5cf9004ac39f442493b90a930022860ae  ./small.zip
$ b2 upload-file --sha1 5f7270e5cf9004ac39f442493b90a930022860ae molomby small.zip small.zip 
small.zip: 100%|█████████████████| 13.4M/13.4M [00:10<00:00, 1.24MB/s]
URL by file name: https://f002.backblazeb2.com/file/molomby/small.zip
URL by fileId: https://f002.backblazeb2.com/b2api/v2/b2_download_file_by_id?fileId=4_z80c47a5d0d48e29c7f100612_f107dea251c0f0d02_d20200518_m032455_c002_v0001126_t0003
{
  "action": "upload", 
  "fileId": "4_z80c47a5d0d48e29c7f100612_f107dea251c0f0d02_d20200518_m032455_c002_v0001126_t0003", 
  "fileName": "small.zip", 
  "size": 13412956, 
  "uploadTimestamp": 1589772295000
}
$ b2 get-file-info 4_z80c47a5d0d48e29c7f100612_f107dea251c0f0d02_d20200518_m032455_c002_v0001126_t0003 | jq '.contentSha1'
"9c3dd5ad33877640b83b7e653cea4bbeec188c8c"
```

Great, we managed to store a checksum!

But wait.. the checksum we specified is different than the one that is stored... `5f7270e5cf9004ac39f442493b90a930022860ae` != `9c3dd5ad33877640b83b7e653cea4bbeec188c8c`. What's going on here?

Turns out, when `b2 upload-file` reads a "small" file from the disk it _calculates it's own hash and ignores the supplied value_ (#478). In this scenario, my crappy [consumer grade HDD or lack of ECC memory](https://github.com/Backblaze/B2_Command_Line_Tool/issues/604#issuecomment-566299914) introduced an error either on the first read (when I manually calculated the checksum) or the second read (when `b2` uploaded the file). If it's the latter, the checksum functionality has **completely failed** in it's job. It's literally been given a file and a checksum then read and uploaded a corrupted copy, ignoring the exact thing intended to prevent this.

Admittedly, this is hypothetical; I haven't observed corruption occurring like this. But it doesn't change the fact the CLI is ignoring supplied checksums and "succeeding" even when they indicate that a serious error has occurred somewhere.

## Not Computed Server-Side

Even if _all_ these problems were fixed, the fact that checksums aren't calculated by the b2 platform for whole large files is problematic.

Say I have some important data -- a single 4 Gb file that I want to store, without it being corrupted (the kind of problem the `b2` platform is trying to solve, no?). What's my workflow to actually achieve this?

I know my consumer hardware is flaky but that's ok, I have 3 copies of the file on different media and a sha1 from the original anyway.

Let's assume all the other problem described in this are fixed and the docs are correct. I can upload this file and set the sha1 checksum as metadata using `--sha1`. As I understand it, the files is split during upload with each chunk being checksummed and stored separately (?). The checksumming here protects the data in transit, but how to I verify the file was _read_ correctly during upload and actually matches the original?

Since the platform itself only considers checksums for the parts, I need to download the entire file and re-compute the checksum to verify the whole stored file is correct. (Please let me know if I'm wrong on this point -- if there's a better way I haven't been able to fine it.)

If I need to re-download everything I upload to check the upload was correct, what's the point in even providing the sha1 to start with?

---

Footnotes:

* Using `b2` CLI v1.4.2
* I'm sure it's possible to work around some of the problems outlined here using the HTTP API. For me, doing so would defeat the purpose of using the CLI at all.
* It's quite possible this is a PEBKAC situation and I'm missing something significant that solves all my problems. If that's the case, it's probably worth taking this as feedback on your docs. I'm a technically minded person, using your platform and tools for (as far as I can tell) the kind of thing they're meant to be used for but I'm consistently hitting problems.

**Comment by ppolewicz** [comment]
Releases of b2sdk 1.1.0 and b2cli v2.0.0 are close. After that, a few things will change. To name a few:
1) large files will get a sha1 calculated client-side by default unless 2) it is provided by the user (so it will actually work)

As for 3), I not a b2 server developer, but maybe something can be done. I need to think about it a little bit.

**Comment by molomby** [comment]
Thanks @ppolewicz, v2 sounds good.

> As for 3), I not a b2 server developer, but maybe something can be done. I need to think about it a little bit.

Re: calculating server-side -- For context, I'm reasonably bandwidth constrained at the moment so this is a real drag. My current workaround is a VPS in US west that I'm using solely to read files back out of b2 and verify. Having a `b2 compute-sha1` command or similar that did basically the same thing (streamed an entire file off the disks and computed a hash) from within the platform itself would be great.

It'd be a comparatively heavy operation I'm sure, but honestly, if this existed as a "Class D" transaction that cost (a small amount of) money for each call, I'd be using it.

**Comment by ppolewicz** [comment]
How large are the files that you are working with?

**Comment by ppolewicz** [comment]
I am not sure if I understand the last part of your issue, checking if the file has been read correctly, the large file issue specifically. For a small file, you'd sync or upload it to the server, b2cli would read it once the file while streaming it out to the server, server would do the same and the checksum would be validated at the end with both ends failing the operation of the checksum doesn't match (so in case there was an issue with transit), but the file is read just once from the drive. You could, before uploading the file, run a checksum in bash before calling `b2 upload-file` and then compare the result of your first read with the result returned by b2cli/server... but the data would probably be read from the actual drive just once, since the kernel of your system would keep the data in a RAM buffer between the two reads (unless you manage to flush it, you read it with some special flag or your kernel decides to evict the buffer between reads for some reason, for example due to memory pressure). I can't see this as an issue specific to large files. Is it for all files, or am I missing something (are you avoiding the kernel buffers to force another read from drive)?

**Comment by molomby** [comment]
> For a small file, you'd sync or upload it to the server, b2cli would read it once the file while streaming it out to the server, server would do the same and the checksum would be validated at the end with both ends failing the operation of the checksum doesn't match (so in case there was an issue with transit), but the file is read just once from the drive.

Ok, that's better than I though at least. I wasn't sure if the server actually performed this verification and couldn't find a way to test it.

> You could, before uploading the file, run a checksum in bash before calling b2 upload-file and then compare the result of your first read with the result returned by b2cli/server... but the data would probably be read from the actual drive just once, since the kernel of your system would keep the data in a RAM buffer between the two reads [...]

I'd still consider it a pretty significant bug that the supplied checksum is ignored (my "Supplied Value Ignored" point above).

Consider a scenario where the user has a "small" file with a _previously computed_ sha1 (eg. a checksum that was computed years ago when the file was archived elsewhere):

```sh
$ ls -lh            
-rw-r--r--  1 molomby  staff    92M  8 Jun  2013 small.zip
-rw-r--r--  1 molomby  staff    52B  8 Jun  2013 small.zip.sha1
$ cat small.zip.sha1 
148e604087ffe6ceff3f1ceabdf95b78bf144b14  small.zip
```

Note, both these files are from 2013 and the file hasn't been read since then. We've no idea the state the storage medium is in.

To ensure the integrity of the file the existing/original checksum should be used whenever possible, right? When uploading the file to b2, my first thought would be to do something like this:

```sh
$ b2 upload-file --sha1 $(cat small.zip.sha1 | cut -f1 -d' ') molomby small.zip small.zip
```

Do you see the problem? If `small.zip` has become corrupted in the last 7 years or the media is unreliable the upload still "succeeds" and completes without error. The **correct checksum provided is discarded** and a _new_ checksum is created from a corrupt copy when the file is read for upload. This _new_ checksum is the one then "verified" server-side, giving the user the incorrect assurance that the file now stored by b2 is the same as the file they created 7 years earlier.

What I'd expect to happen here is that my _original_ checksum is used to verify the file being read and sent to the server to be verified against the copy stored. This would cause the upload attempt above to error out, prompting the user to either source another copy of the file or at least retry the operation, hoping to get an uncorrupted read (clean the CD, instigate some kind of data recovery, etc.).

Basically, **the sha1 supplied by the user should always be considered the source of truth** and should take precedence as the correct value to check against.

**Comment by ppolewicz** [comment]
I agree, but I think that's a duplicate of https://github.com/Backblaze/B2_Command_Line_Tool/issues/604. Are there any other issues that you are experiencing?

**Comment by ppolewicz** [comment]
A month ago, I said (and @molomby upvoted):
> 1. large files will get a sha1 calculated client-side by default unless 2) it is provided by the user (so it will actually work)

This is not the best way. It doubles the io cost and hashing cost. When I wrote this, I thought this is the only way - but actually B2 supports S3 api now and there we can retrieve information about original parts, their ranges and checksums (md5 I think), so using that we will be able to create a `b2 verify-file file_id local_path`, which will not download any data from the cloud and will even be able to tell which part is damaged... But the upload cost does not get hit with double IO and double hashing, so this is preferable. Unfortunately we can't fit this in b2cli 2.0.0 release, which is any day now.

**Comment by ppolewicz** [comment]
Improvements for sha1 handling were done in https://github.com/Backblaze/b2-sdk-python/pull/138

**Comment by ppolewicz** [comment]
Improvements for sha1 handling were done in https://github.com/Backblaze/b2-sdk-python/pull/138
